architecture/2024-08-bicepdeployment-controller.md

```diff
@@ -0,0 +1,693 @@
+  // Template is the ARM JSON template that defines the resources to deploy.
+  Template string `json:"template"`
+
+  // Parameters is the ARM JSON parameters file that specifies the parameters for the template.
+  Parameters string `json:"parameters"`
+}
+
+// BicepDeploymentStatus defines the observed state of the Bicep Deployment.
+type BicepDeploymentStatus struct {
+  // ObservedGeneration is the most recent generation observed for this BicepDeployment.
+  ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Scope is the resource ID of the scope.
+	Scope string `json:"scope,omitempty"`
+
+	// Resource is the resource ID of the deployment.
+	Resource string `json:"resource,omitempty"`
+
+	// Operation tracks the status of an in-progress provisioning operation.
+	Operation *ResourceOperation `json:"operation,omitempty"`
+
+	// Phrase indicates the current status of the Bicep Deployment.
+	Phrase BicepDeploymentPhrase `json:"phrase,omitempty"`
+
+  // Description is a human-readable description of the status of the Bicep Deployment.
+  Description string `json:"description,omitempty"`
+
+  // Resources is a list of resources that are created by the BicepDeployment.
+  Resources []BicepDeploymentResourceRef `json:"resources,omitempty"`
+}
+
+// BicepDeploymentPhrase is a string representation of the current status of a Bicep Deployment.
+type BicepDeploymentPhrase string
+
+const (
+	// BicepDeploymentPhraseUpdating indicates that the Bicep Deployment is being updated.
+	BicepDeploymentPhraseUpdating BicepDeploymentPhrase = "Updating"
+
+	// BicepDeploymentPhraseReady indicates that the Bicep Deployment is ready.
+	BicepDeploymentPhraseReady BicepDeploymentPhrase = "Ready"
+
+	// BicepDeploymentPhraseFailed indicates that the Bicep Deployment has failed.
+	BicepDeploymentPhraseFailed BicepDeploymentPhrase = "Failed"
+
+	// BicepDeploymentPhraseDeleting indicates that the Bicep Deployment is being deleted.
+	BicepDeploymentPhraseDeleting BicepDeploymentPhrase = "Deleting"
+
+	// BicepDeploymentPhraseDeleted indicates that the Bicep Deployment has been deleted.
+	BicepDeploymentPhraseDeleted BicepDeploymentPhrase = "Deleted"
+)
+```
+
+#### BicepDeploymentResource Custom Resource Definition
+
+The `BicepDeploymentResource` CRD is another CRD that will be responsible for tracking the state of individual resources. The CRD will have the following fields:
+
+```go
+// BicepDeploymentResourceSpec defines the desired state of a BicepDeployment
+type BicepDeploymentResourceSpec struct {
+  // ResourceID is the resource ID of the resource to deploy.
+  ResourceID string `json:"resourceID"`
+}
+
+type BicepDeploymentResourceStatus struct {
+  // ObservedGeneration is the most recent generation observed for this BicepDeploymentResource.
+  ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+  // Operation tracks the status of an in-progress provisioning operation.
+  Operation *ResourceOperation `json:"operation,omitempty"`
```

Comment:
This is how we will understand if the operation is successful or not or if it is still going on, I believe.